### ISSUE TITLE
pretraining code

### ISSUE THREAD

**Issue by mindsetmind** [THUDM/OAG-BERT#1]
Thanks for the nice work.  Would the pretraining code be released?

**Comment by Somefive** [comment]
Currently, we do not have a plan to release the pretraining code since we are still exploring more based on that. Sorry for your request. We might release the pretraining code in the future. The first pretraining stage mainly follows the BERT pretraining code while the second stage is associated with some customized masking codes. We recommend you try the code in [DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/master/docs/_tutorials/bert-pretraining.md).

**Comment by mindsetmind** [comment]
Thanks for the prompt reply and recommeded deepspeed project. Also I am quite interested in the implementation of customized masking in the second stage. Could you also please share some resources for reference when you implement the masking strategies?

**Comment by Somefive** [comment]
It might be difficult to share with you the resource of implementing masking strategies as we developed our own masking strategy without any code reference (of course we borrow some ideas from other papers).
Generally, there are two ways of doing masking. One is static masking. In this way, you can change the pre-processing code and generate masks before running the pretraining scripts. The other way is dynamic masking which means you need to change the DataLoader and generate masks during the pretraining stage. The first way is faster while the second way is more flexible. However, due to some DataLoader design problem, the second way is vulnerable to memory leaks which means you may encounter a increasing usage of memory if the code is not well written. 

Some of the DeepSpeed references are listed below.
1. [DeepSpeed Official Tutorial](https://www.deepspeed.ai/tutorials/bert-pretraining/)
2. [Pytorch Code of BERT Pretraining with DeepSpeed](https://github.com/microsoft/DeepSpeedExamples/tree/master/bing_bert)
3. [The DataLoader in DeepSpeed which you can modify to implement dynamic masking](https://github.com/microsoft/DeepSpeedExamples/blob/73fbeb9167b92fdec2e88bd6b143b6b6dc7cd61c/bing_bert/turing/dataset.py#L295)

**Comment by Somefive** [comment]
As for the masks themselves, the meta-info masks in the paper are mainly based on static masking. However, for other later versions, we switch to dynamic masking. 
For meta-info masks, we first calculate the entities except for the pure text (title+abstract) and do span masking like SpanBERT, then we calculate the remaining masking quota to mask the texts as the original BERT does.

**Comment by ccclyu** [comment]
Thanks greatly for your suggestions!  I would follow the above and try to implement them!

**Comment by mindsetmind** [comment]
Thanks a lot! I would have a deep look of DeepSpeed codes and your suggestions!
